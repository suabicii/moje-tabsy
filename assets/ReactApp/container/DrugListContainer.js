import React, {useState} from "react";
import {createContainer} from "unstated-next";

const useDrugListing = (initialState = []) => {
    const [drugList, setDrugList] = useState(initialState);
    const addDrug = ({id, name, quantity, quantityMax, unit, dosing, dosingMoments}) => setDrugList([
        ...drugList,
        {
            id,
            name,
            quantity,
            quantityMax,
            unit,
            dosing,
            dosingMoments
        }
    ]);
    const removeDrug = id => setDrugList(drugList.filter(drug => drug.id !== id));
    const editDrug = (id, updates = {}) => setDrugList(drugList.map(drug => drug.id === id ? {...drug, ...updates} : drug));
    return {drugList, addDrug, removeDrug, editDrug};
};

export const DrugListContainer = createContainer(useDrugListing);