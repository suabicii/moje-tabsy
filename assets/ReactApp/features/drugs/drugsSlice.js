import {createSelector, createSlice} from "@reduxjs/toolkit";
import {fetchData} from "../../utils/fetchData";

const drugsSlice = createSlice({
    name: 'drugs',
    initialState: [],
    reducers: {
        addDrug: (state, action) => {
            const {id, name, quantity, quantityMax, unit, dosing, dosingMoments} = action.payload;
            state.push({id, name, quantity, quantityMax, unit, dosing, dosingMoments});
        },
        removeDrug: (state, action) => {
            const id = action.payload;
            return state.filter(drug => drug.id !== id);
        },
        editDrug: (state, action) => {
            const {id, ...updates} = action.payload;
            const drugIndex = state.findIndex(drug => drug.id === id);
            if (drugIndex !== -1) {
                state[drugIndex] = {...state[drugIndex], ...updates};
            }
        },
        setDrugs: (state, action) => action.payload
    },
    sort: true
});

const {reducer, actions} = drugsSlice;
export const {addDrug, removeDrug, editDrug, setDrugs} = actions;

const defaultDrugsSelector = state => state.drugs;

export const sortedDrugsSelector = createSelector(
    defaultDrugsSelector,
    drugs => [...drugs].sort((a, b) => a.name.localeCompare(b.name))
);

export const fetchDrugs = () => async dispatch => {
    await fetchData('/api/drug-list').then(data => {
        dispatch(setDrugs(data));
    }).catch(err => {
        console.log(err);
    });
};

export default reducer;