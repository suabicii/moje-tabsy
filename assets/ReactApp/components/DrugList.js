import React, {useEffect, useState} from "react";
import {mainRoute} from "../routers/AppRouter";
import {useNavigate} from "react-router-dom";
import {DrugListContainer} from "../container/DrugListContainer";
import DrugForm from "./forms/DrugForm";

function DrugList(props) {
    const navigate = useNavigate();
    const {drugList, removeDrug} = DrugListContainer.useContainer();
    const [isEditFormVisible, setIsEditFormVisible] = useState(false);
    const [isAddFormVisible, setIsAddFormVisible] = useState(false);
    const [drugDataForEdit, setDrugDataForEdit] = useState(undefined);

    useEffect(() => {
        localStorage.setItem('drugs', JSON.stringify(drugList));
    });

    return (
        <>
            <div className="card text-center mt-3">
                <div className="card-header">
                    Leki/suplementy <i className="fa-solid fa-pills"></i>
                </div>
                <div className="card-body">
                    <h5 className="card-title">Lista leków i suplementów:</h5>
                    <ul className="card-text list-group">
                        {drugList.map(drug => <li key={drug.id} className="list-group-item" data-testid="drug">
                                <strong>{drug.name}</strong>: {drug.quantity}/{drug.quantityMax} {`${drug.unit}, `}
                                <strong>Dzienna
                                    dawka: </strong> {drug.dosing} {drug.unit} {`${Object.keys(drug.dosingMoments).length} raz(-y) dziennie `}
                                {
                                    // Delete button
                                    props.isEditMode &&
                                    <button className="btn btn-danger rounded-circle float-md-end"
                                            data-testid={`remove-drug-${drug.id}`} onClick={() => {
                                        removeDrug(drug.id);
                                    }}>
                                        <i className="fa-solid fa-trash-can"></i>
                                    </button>
                                }
                                {
                                    // Edit button
                                    props.isEditMode &&
                                    <button className="btn btn-info rounded-circle float-md-end"
                                            data-testid={`edit-drug-${drug.id}`} onClick={() => {
                                        setIsEditFormVisible(true);
                                        setDrugDataForEdit(drug);
                                    }}>
                                        <i className="fa-solid fa-pencil"></i>
                                    </button>
                                }
                            </li>
                        )}
                    </ul>
                    <div className="d-grid">
                        <button className="btn btn-primary mt-2" data-testid="add-drug" onClick={() => {
                            if (props.isEditMode) {
                                setIsAddFormVisible(true);
                            } else {
                                navigate(`${mainRoute}/drug-list`);
                            }
                        }}>
                            {props.isEditMode ? 'Dodaj ' : 'Edytuj '}
                            {props.isEditMode ? <i className="fa-solid fa-plus"></i> :
                                <i className="fa-solid fa-pencil"></i>}
                        </button>
                    </div>
                </div>
            </div>
            {
                props.isEditMode &&
                <>
                    {
                        isAddFormVisible &&
                        <div className="card mt-3">
                            <DrugForm setIsFormVisible={setIsAddFormVisible}/>
                        </div>
                    }
                    {
                        isEditFormVisible &&
                        <div className="card mt-3">
                            <DrugForm drug={drugDataForEdit} setIsFormVisible={setIsEditFormVisible}/>
                        </div>
                    }
                </>
            }
        </>
    );
}

export default DrugList;