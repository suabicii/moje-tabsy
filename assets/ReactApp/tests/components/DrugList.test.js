/**
 * @jest-environment jsdom
 * */
import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";
import DrugList from "../../components/DrugList";
import {BrowserRouter} from "react-router-dom";
import {fireEvent, render, screen} from "@testing-library/react";
import {DrugListContainer} from "../../container/DrugListContainer";
import drugs from "./fixtures/drugs";
import {act} from "react-dom/test-utils";

const renderDrugList = () => {
    const root = document.createElement('div');
    document.body.appendChild(root);
    render(
        <DrugListContainer.Provider initialState={drugs}>
            <BrowserRouter>
                <DrugList isEditMode={true} customRoot={root}/>
            </BrowserRouter>
        </DrugListContainer.Provider>
    );
};

it('should correctly render DrugList', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(
        <BrowserRouter>
            <DrugList isEditMode={true}/>
        </BrowserRouter>
    );

    expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should render add drug form after clicking add button', () => {
    renderDrugList();
    fireEvent.click(screen.getByTestId('add-drug'));
    expect(screen.getByRole('form')).toBeVisible();
});

it('should render edit drug form after clicking edit button', () => {
    renderDrugList();
    // click edit button near first list item
    fireEvent.click(screen.getByTestId('edit-drug-1')); // from data-testid, 1 is id of first item
    expect(screen.getByRole('form')).toBeVisible();
});

it('should delete drug from list after clicking delete button',  async () => {
    await act(() => {
        renderDrugList();
    });
    const drugListLengthBeforeDelete = screen.getAllByTestId('drug').length;

    await act(async () => {
        fireEvent.click(screen.getByTestId('remove-drug-1'));
    });
    const drugListLengthAfterDelete = screen.getAllByTestId('drug').length;

    expect(drugListLengthAfterDelete).toBeLessThan(drugListLengthBeforeDelete);
});