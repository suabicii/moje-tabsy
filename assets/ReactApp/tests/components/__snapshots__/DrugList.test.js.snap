// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render DrugList 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <DrugList
    isEditMode={true}
  />
</Router>
`;
